### PR TITLE
Pin airbyte tools to .version source-of-truth files, add to CI configs, improve check_requirements.sh, and build.gradle for dockerBuild tasks

### PR DESCRIPTION
**Pull request by noahkawasakigoogle**
## What

Create single source of truth `.version` files for various things Airbyte uses for development. This was inspired from seeing https://github.com/airbytehq/airbyte/issues/3028 and filing https://github.com/airbytehq/airbyte-internal-issues/issues/363. Refactors the dependent github workflow yaml files, `tools/bin/check_requirements.sh` script, and `build.gradle` so all dependent processes in the repo get their versions from the source of truth files instead of hardcoded values. This ensures that, across files, languages, and tools in this repo, everything always gets the same value for a given dependency to use. 

## How
*Describe the solution*

Create `.version` files for: `.node-version`, `.pip-version`, `.java-version` to act as _the_ version value to use in the repo. Refactor github workflow scripts to read this file as a step in the job, and set the relevant env variable in the job definition (`JAVA_VERSION`, `PIP_VERSION`, `PYTHON_VERSION`), and then use this value in the `setup-java` github action steps.

It works by first creating a step which writes the env variable into `GITHUB_ENV`:

```
      - name: Set java-version
        run: |
          echo "JAVA_VERSION="$(cat ../../.java-version)"" >> $GITHUB_ENV
```

And then using that value downstream:

```
        with:
          java-version: "${{ env.JAVA_VERSION }}"
```

**NOTE**: I do not have access to Airbyte CI, so I cannot validate those workflow changes^ if someone from Airbyte can run them I can make necessary changes. Also, there was no DRYer way to accomplish this that I could find from Github Actions, the repetitive set java-version steps are necessary because each job runs on a separate node and thus cannot share environment variables across.

This makes it very very easy to upgrade tools in the future, see this [PR](https://github.com/airbytehq/airbyte/pull/7964/files#diff-1aab6d4260f220e37df7b9118875cc3aabd07dcf9651c23c63cc3af6c59b47c3L49) for an example where there is potential to miss certain files from searching the repo. Now all someone needs to do is change the `.java-version` file and that change will cascade through everything. 

I've also changed the root level `build.gradle` file to read `jdkVersion` from `.java-version` so the docker builds get their `JDK_VERSION` build arg from that file as well

## Recommended reading order
1. Any order, there is no dependencies between any files for the sake of understanding the changes. They are pretty straightforward and you can view them in the github UI order

## 🚨 User Impact 🚨
None

## Pre-merge Checklist
None applicable

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/9230)

### CODE REVIEW

File: tools/bin/check_requirements.sh [hunk 1]
```diff
@@ -1,68 +1,102 @@
...
-        if [[ "${version}" > "${desired}" || "${version}" == "${desired}" ]];
+        java_version="${array[1]}"
+        printf "${java_version} installed"
+        if [[ "${java_version}" == "${DESIRED_JAVA_VERSION}" ]];
             then
-                printf " and functional"
+                printf " and matches version ${DESIRED_JAVA_VERSION}"
             else
-                printf " but not functional, must have version ${desired} at least"
+                printf " but does not match version ${DESIRED_JAVA_VERSION}, you might see unexpected behavior"
```
Comment (noahkawasakigoogle) [inline tools/bin/check_requirements.sh:30]: Changed the logic here to require the exact version. Using a higher version could have issues and it should not be "okay" if someone has a higher version than the required version even if it is better than having a lower version.

File: tools/bin/check_requirements.sh [hunk 2]
```diff
@@ -1,68 +1,102 @@
...
-                printf " but not functional, must have version ${desired} at least"
+                printf " but does not match version ${DESIRED_PYTHON_VERSION}, you might see unexpected behavior"
         fi
     else
-        printf "not installed, must have version ${desired} at least"
+        printf "not installed, please install Python ${DESIRED_PYTHON_VERSION}"
 fi;
+
+printf "\n";
+DESIRED_NODE_VERSION="$(cat ../../.node-version)"
```
Comment (noahkawasakigoogle) [inline tools/bin/check_requirements.sh:77]: Added Node checking

**Comment by noahkawasakigoogle** [comment]
FYI I need to get an approval from my employer (Google) on signing the CLA first. They've got some processes around open source contributions

File: .github/workflows/detect-changed-modules-and-build-reports.yml [hunk 3]
```diff
@@ -13,10 +13,12 @@ jobs:
         uses: actions/checkout@v2
         with:
           fetch-depth: 1000
+      - name: Set python-version
+        run: echo "PYTHON_VERSION="$(cat ../../.python-version)"" >> $GITHUB_ENV
```
Comment (noahkawasakigoogle) [inline .github/workflows/detect-changed-modules-and-build-reports.yml:17]: This relative path might just need to be `.python-version`. Not sure what directory github workflows run out of

**Review by jrhizor — APPROVED** [review]
Nice! Thanks!

**Comment by noahkawasakigoogle** [comment]
Some files in this PR are not MIT licensed, so I cannot make this change :(

**Comment by bleonard** [comment]
We'll want to have another spot to record/share the docker situation as well.

`python:3.9-slim` https://github.com/airbytehq/airbyte/blob/5c8bd3a46def819c46fce60c103e376d04948b3e/airbyte-integrations/connectors/source-file/Dockerfile#L1

`FROM python:3.9.11-alpine3.15 as base` https://github.com/airbytehq/airbyte/blob/5c8bd3a46def819c46fce60c103e376d04948b3e/airbyte-integrations/bases/source-acceptance-test/Dockerfile

`python-version: 3.9` https://github.com/airbytehq/airbyte/blob/5c8bd3a46def819c46fce60c103e376d04948b3e/.github/actions/ci-tests-runner/action.yml#L32

`python_requires=">=3.9"` https://github.com/airbytehq/airbyte/blob/5c8bd3a46def819c46fce60c103e376d04948b3e/tools/ci_code_validator/setup.py#L26

Or anywhere else from the recent PR https://github.com/airbytehq/airbyte/pull/11763/files

**Pull request closed without merging**